src/pool/srv_pool.c

```diff
@@ -856,25 +841,58 @@ ds_pool_enable_exclude(void)
 	pool_disable_exclude = false;
 }
 
+static void
+pool_exclude_rank_ult(void *data)
+{
+	struct ds_pool_exclude_arg     *arg = data;
+	daos_prop_t		prop = { 0 };
+	struct daos_prop_entry	*entry;
+	struct pool_svc		*svc;
+	int			rc = 0;
+
+	pool_svc_get_leader(arg->svc);
+	svc = arg->svc;
+	rc = ds_pool_iv_prop_fetch(svc->ps_pool, &prop);
+	if (rc)
+		D_GOTO(free, rc);
+
+	entry = daos_prop_entry_get(&prop, DAOS_PROP_PO_SELF_HEAL);
+	D_ASSERT(entry != NULL);
+	if (!(entry->dpe_val & DAOS_SELF_HEAL_AUTO_EXCLUDE)) {
+		D_DEBUG(DB_MGMT, "self healing is disabled\n");
+		D_GOTO(prop_fini, rc);
+	}
+
+	rc = pool_svc_exclude_rank(svc, arg->rank);
+	D_DEBUG(DB_MGMT, DF_UUID" exclude rank %u : rc %d\n",
+		DP_UUID(svc->ps_uuid), arg->rank, rc);
+
+prop_fini:
+	daos_prop_fini(&prop);
+free:
+	if (rc)
+		D_ERROR("pool "DF_UUID" exclud failed: rc %d\n",
+			DP_UUID(svc->ps_uuid), rc);
+	pool_svc_put_leader(arg->svc);
+	D_FREE_PTR(arg);
+}
+
 static int
 pool_exclude_rank(struct pool_svc *svc, d_rank_t rank)
 {
-	struct ds_pool_exclude_arg	*ult_arg;
-	int				rc;
+	struct ds_pool_exclude_arg      *ult_arg;
+	int                             rc;
 
 	D_ALLOC_PTR(ult_arg);
 	if (ult_arg == NULL)
 		D_GOTO(out, rc = -DER_NOMEM);
 
-	pool_svc_get_leader(svc);
```

Comment:
I'm afraid dropping this is not safe, for the newly-created ULT may not be scheduled until svc is destroyed for some reason. I will work out a patch for you to merge into this PR.

Reply:
done

Reply:
Hmm, this is not resolved yet. svc reference counters can't be modified from another xstream. I'm working on a patch for you to merge into this PR.

Reply:
(And most svc methods can't be called from another xstream too.)

Reply:
Great. Thanks!